File: 2024/day_24/solution.js
function isTreesSynchronized(tree1, tree2) {
  function mirror(t1, t2) {
    if (!t1 && !t2) return true;
    if (!t1 || !t2) return false;

    return (
      t1.value === t2.value &&
      mirror(t1.left, t2.right) &&
      mirror(t1.right, t2.left)
    );
  }

  return [mirror(tree1, tree2), tree1?.value];
}

const tree1 = {
  value: "🎄",
  left: { value: "⭐" },
  right: { value: "🎅" },
};

const tree2 = {
  value: "🎄",
  left: { value: "🎅" },
  right: { value: "⭐" },
};

console.log(isTreesSynchronized(tree1, tree2)); // [true, '🎄']

const tree3 = {
  value: "🎄",
  left: { value: "🎅" },
  right: { value: "🎁" },
};

console.log(isTreesSynchronized(tree1, tree3)); // [false, '🎄']

const tree4 = {
  value: "🎄",
  left: { value: "⭐" },
  right: { value: "🎅" },
};

console.log(isTreesSynchronized(tree1, tree4)); // [false, '🎄']

console.log(isTreesSynchronized({ value: "🎅" }, { value: "🧑‍🎄" })); // [false, '🎅']

File: 2024/day_24/solution.py
def is_trees_synchronized(tree1, tree2):
    def are_mirrors(node1, node2):
        # Base case: Both nodes are None
        if not node1 and not node2:
            return True
        
        # One of the nodes is None, but not both
        if not node1 or not node2:
            return False
        
        # Check if values match and recursively check left/right subtrees as mirrors
        return (
            node1["value"] == node2["value"]
            and are_mirrors(node1.get("left"), node2.get("right"))
            and are_mirrors(node1.get("right"), node2.get("left"))
        )

    # The root value of tree1
    root_value = tree1["value"] if tree1 else None

    # Check if the trees are mirrors
    is_synchronized = are_mirrors(tree1, tree2)

    return [is_synchronized, root_value]
